unixfs/src/walk.rs

```diff
@@ -0,0 +1,1040 @@
+        let mut path = PathBuf::new();
+        path.push(name);
+        Self {
+            kind: InnerKind::Symlink(None),
+            path,
+            metadata,
+            depth,
+        }
+    }
+
+    pub fn as_entry(&self) -> Entry<'_> {
+        use InnerKind::*;
+        match &self.kind {
+            RootDirectory | BucketAtRoot => Entry::RootDirectory(&self.path, &self.metadata),
+            RootBucket(cid) => Entry::Directory(cid, &self.path, &self.metadata),
+            Bucket(cid) => Entry::Bucket(cid, &self.path),
+            Directory(cid) => Entry::Directory(cid, &self.path, &self.metadata),
+            File(cid, _, sz) => Entry::File(cid.as_ref(), &self.path, &self.metadata, *sz),
+            Symlink(cid) => Entry::Symlink(cid.as_ref(), &self.path, &self.metadata),
+        }
+    }
+
+    fn set_path(&mut self, name: &str, depth: usize) {
+        while self.depth >= depth && self.depth > 0 {
+            assert!(self.path.pop());
+            self.depth -= 1;
+        }
+
+        self.path.push(name);
+        self.depth = depth;
+    }
+
+    fn as_directory(&mut self, cid: Cid, name: &str, depth: usize, metadata: Metadata) {
+        use InnerKind::*;
+        match self.kind {
+            RootDirectory
+            | BucketAtRoot
+            | Bucket(_)
+            | RootBucket(_)
+            | Directory(_)
+            | File(_, None, _)
+            | Symlink(_) => {
+                self.kind = Directory(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            }
+            ref x => todo!("dir after {:?}", x),
+        }
+    }
+
+    fn as_bucket_root(&mut self, cid: Cid, name: &str, depth: usize, metadata: Metadata) {
+        use InnerKind::*;
+        match self.kind {
+            RootDirectory
+            | BucketAtRoot
+            | Bucket(_)
+            | RootBucket(_)
+            | Directory(_)
+            | File(_, None, _)
+            | Symlink(_) => {
+                self.kind = RootBucket(cid);
+                self.set_path(name, depth);
+                self.metadata = metadata;
+            }
+            ref x => todo!("root bucket after {:?}", x),
+        }
+    }
+
+    fn as_bucket(&mut self, cid: Cid, name: &str, depth: usize) {
+        use InnerKind::*;
+        match self.kind {
+            BucketAtRoot => {
+                assert_eq!(self.depth, depth, "{:?}", self.path);
+            }
+            RootBucket(_) | Bucket(_) | File(_, None, _) | Symlink(_) => {
+                self.kind = Bucket(cid);
+
+                if name.is_empty() {
+                    // continuation bucket going bucket -> bucket
+                    while self.depth > depth {
+                        assert!(self.path.pop());
+                        self.depth -= 1;
+                    }
+                } else {
+                    self.set_path(name, depth);
+                }
```

Comment:
this should be `assert!(name.is_empty());` and the `else` path removed.

I think I'll use `self.depth = self.depth.checked_sub(1).expect("where");` in all of these cases.